src/index.ts

```diff
@@ -48,6 +51,120 @@ const observeGitInfo = async (
   return () => clearTimeout(timer);
 };
 
+const getConfig = (
+  task: TaskName,
+  previousBuildProgress?: RunningBuildPayload["previousBuildProgress"]
+) => {
+  if (!isKnownStep(task)) throw new Error(`Unknown step: ${task}`);
+  const stepDurations = BUILD_STEP_ORDER.map((step) => {
+    const { startedAt, completedAt } = previousBuildProgress?.[step] || {};
+    return startedAt && completedAt
+      ? completedAt - startedAt
+      : BUILD_STEP_CONFIG[step].estimateDuration;
+  });
+  const totalDuration = stepDurations.reduce((sum, duration) => sum + duration, 0);
+
+  const stepIndex = BUILD_STEP_ORDER.indexOf(task);
+  const startTime = stepDurations.slice(0, stepIndex).reduce((sum, duration) => sum + duration, 0);
+  const endTime = startTime + stepDurations[stepIndex];
+
+  const startPercentage = (startTime / totalDuration) * 100;
+  const endPercentage = (endTime / totalDuration) * 100;
+  return {
+    ...BUILD_STEP_CONFIG[task],
+    startPercentage,
+    endPercentage,
+    stepPercentage: endPercentage - startPercentage,
+  };
+};
+
+const onStartOrProgress =
+  (runningBuildState: ReturnType<typeof useAddonState<RunningBuildPayload>>) =>
+  ({ ...ctx }: Context, { progress, total }: { progress?: number; total?: number } = {}) => {
+    if (isKnownStep(ctx.task)) {
+      const { buildProgressPercentage, stepProgress, previousBuildProgress } =
+        runningBuildState.value;
+
+      const { startPercentage, endPercentage, stepPercentage } = getConfig(
+        ctx.task,
+        previousBuildProgress
+      );
+
+      let newPercentage = startPercentage;
+      if (progress && total) {
+        newPercentage += stepPercentage * (progress / total);
+      }
+
+      if (!["upload", "snapshot"].includes(ctx.task)) {
```

Comment:
```suggestion
      // If the step doesn't have a progress event, simulate one by synthetically updating progress
      // every CONSTANT_NAME ms
      if (!["upload", "snapshot"].includes(ctx.task)) {
```

Reply:
Also, perhaps we should add `hasProgressEvent` to `buildSteps.ts` rather than hardcoding it here?